Use entity_class 'safety' in synology_dsm storage sensors

Mariusthvdb opened the pull request:
and more meaningful icons than 'mdi:test-tube', not sure if these are even necessary, given the device_class that defines the icons too?

this is a follow-up on https://github.com/home-assistant/core/pull/39310#issuecomment-688282030

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Mariusthvdb committed:
Use entity_class: 'safety' in storage sensors

and more meaningful icons then 'mdi:test-tube', not sure if these are even necessary, given the device_class that defines the icons too?

frenck commented on homeassistant/components/synology_dsm/const.py:
```diff
@@ -41,15 +41,15 @@
         ENTITY_UNIT: None,
-        ENTITY_ICON: "mdi:test-tube",
-        ENTITY_CLASS: None,
+        ENTITY_ICON: "mdi:checkbox-marked-circle-outline",
+        ENTITY_CLASS: "safety",
```
Either set an icon or device class, not both.

Mariusthvdb replied:
ok, l've taken out the icons.

please have a look at my comment, the temp sensors have an icon And device_class set too..

frenck replied:
Yes in general these should not be combined. When setting a device class, it is up to the frontend to provide for icons.

Quentame replied:
Other sensors fixed in #39419

Mariusthvdb replied:
> Yes in general these should not be combined. When setting a device class, it is up to the frontend to provide for icons.

cool, changed accordingly. Only Status sensor is still to be reviewed? Has a different state though: `'normal'` ... Not sure if we can set it to d_c 'Problem' ?

Mariusthvdb committed:
Set device_class, not icon

Still, the temperature sensors have both set, should I take these out in this go too? While we're at it....

frenck reviewed (approved):
Thanks, @Mariusthvdb 👍

Quentame commented on homeassistant/components/synology_dsm/const.py:
```diff
@@ -41,15 +41,15 @@
-        ENTITY_ICON: "mdi:test-tube",
-        ENTITY_CLASS: None,
+        ENTITY_ICON: None,
+        ENTITY_CLASS: "safety",
```
from homeassistant.cons import DEVICE_CLASS_TEMPERATURE

Quentame commented:
Also, remove `DEFAULT_DEVICE_CLASS` and the `device_class` property in sensor.py added in #39310.

Quentame reviewed (changes requested):
Also, remove `DEFAULT_DEVICE_CLASS` and the `device_class` property in sensor.py added in #39310.

Mariusthvdb committed:
added device_class temperature to temp sensors

and removed explicit icons
do we need to set a D_c also on the Status sensor? line 187

Mariusthvdb commented:
> Also, remove `DEFAULT_DEVICE_CLASS` and the `device_class` property in sensor.py added in #39310.

just to be sure: you want me to undo all of the changes in this PR: https://github.com/home-assistant/core/commit/87425d4ab1b7be00c9405bc63fe912230d838f7c#diff-c2e829719830d265736e49e94d56b0c0

Mariusthvdb committed:
revert device_class Safety

which is now set in const.py
following up on https://github.com/home-assistant/core/pull/39757#pullrequestreview-483705147

Quentame commented:
> > Also, remove `DEFAULT_DEVICE_CLASS` and the `device_class` property in sensor.py added in #39310.
> 
> just to be sure: you want me to undo all of the changes in this PR: [87425d4#diff-c2e829719830d265736e49e94d56b0c0](https://github.com/home-assistant/core/commit/87425d4ab1b7be00c9405bc63fe912230d838f7c#diff-c2e829719830d265736e49e94d56b0c0)

Yep because the device_class is already set up with the SynologyDSMEntity and the const.py

Quentame reviewed (commented):
Can you revert this ?

Already included in #39419, this will avoid conflict.

Mariusthvdb commented:
this is getting somewhat confusing, sorry.
Ive reverted the device_class on the storage binary sensor, and will now revert the temp device_class. Hope to fulfill all that is asked?

Quentame commented:
> please let me know exactly what you want reverted (or feel free to do so yourself of course)

Whooops, I though I commented on the commit

Quentame committed:
Use DEVICE_CLASS_SAFETY from const

+ revert temp to avoid conflict

Quentame reviewed (commented):
Nice now, will test before approval.

Quentame commented:
OK so the `safety` device class is not on the HA const

Maybe we should add it here https://github.com/home-assistant/core/blob/973524a4e68efee4c1e484e5c726d10f656a6db1/homeassistant/const.py#L214-L227

What do you think @frenck ?

frenck commented:
You need to import `DEVICE_CLASS_SAFETY` from the matching platform, in this case `binary_sensor`

`from homeassistant.components.binary_sensor import DEVICE_CLASS_SAFETY`

Quentame committed:
const from binary

Mariusthvdb commented:
so are we good now here? The 1 change requested was fulfilled already @Quentame ?

Quentame commented on homeassistant/components/synology_dsm/const.py:
```diff
@@ -227,15 +228,15 @@
     f"{SynoStorage.API_KEY}:disk_smart_status": {
```
`f"{SynoStorage.API_KEY}:disk_smart_status"` and `f"{SynoStorage.API_KEY}:disk_status"` are sensor not binary_sensor

Mariusthvdb replied:
ok, so... what do we need to go on?  had approval 2 days ago, until you came along and things started to become difficulty ;-) Are you saying these can't have the device_class set? which was the point for the beginning of course...

Quentame replied:
I understand that is difficult for you, I should review everything one time, and things that can be nice for people who don't know the code, but is not in real.

I'm saying that this should be binary_sensor not sensor, so there is not only the `const.py` to modify.

Mariusthvdb replied:
ok, but that's another issue altogether? shouldn't we first do this, and after wards see if the team agrees on changing from binary to sensor (which I do agree on btw)
otherwise this PR is diverted too much I fear

Quentame replied:
See 1 line up

Quentame reviewed (commented):
I think the `problem` device class would be better for `SynoStorage` binary sensors.

See : https://developers.home-assistant.io/docs/core/entity/binary-sensor/#available-device-classes

Quentame commented on homeassistant/components/synology_dsm/binary_sensor.py:
```diff
@@ -76,8 +71,3 @@ def is_on(self) -> bool:
         if attr is None:
```
Maybe `getattr(self._api.storage, self.entity_type)(self._device_id) != "normal"`, I need to confirm that.

Storage (binary) sensors deos not return a bool but a status that we have to check.

Otherwise the state will always be `on` !

Quentame replied:
See 1 line up

Mariusthvdb commented:
> I think the `problem` device class would be better for `SynoStorage` binary sensors.
> 
> See : https://developers.home-assistant.io/docs/core/entity/binary-sensor/#available-device-classes

Yes, but as stated in the beginning of this PR, the idea was to have these (binary-) sensors have more meaningful icons than test-tube.... and, going along, to do that, the device_class was attached to them, instead of hard-coding the icons.

So, again, I am fully open to which device_class it should be, as long as we do it as efficient as possible, and of course according to HA guidelines.

as for the underlying code, please let me leave that to you, and I'll follow along, learning what you're doing. 
I was happy with the change in device_class and add meaningful icons to these classes, and setting the class to the sensors. If you feel it is necessary to change the underlying code now, I am lost in that

Please do that in a separate transition? Or this should be closed, which would be a pity really ;-(

Quentame commented:
@Mariusthvdb I need 1 day more to test properly, I'll be busy tonight, sorry.

Put the `device_class` in the const is a good idea thanks :clap: , but I'm pretty sure Storage API status does not return a `bool`. So I'm afraid the binary sensor will always display one state forever.

Good work, still :blush:

Mariusthvdb commented:
Cool, I see that, and yes, it would be great if you would. As Ive noted earlier, we're not used to the state 'Normal' and binaries wont handle that, so if that could be re-formed to a true binary, that would make life much easier. I'll await your findings then.

Quentame commented:
> as for the underlying code, please let me leave that to you, and I'll follow along, learning what you're doing.
> I was happy with the change in device_class and add meaningful icons to these classes, and setting the class to the sensors. If you feel it is necessary to change the underlying code now, I am lost in that
> 
> Please do that in a separate transition? Or this should be closed, which would be a pity really ;-(

OK, we could merge this and I can do code the rest in a separate PR, and tag you.
Is it good to you ?

MartinHjelmare commented on homeassistant/components/synology_dsm/const.py:
```diff
@@ -227,15 +228,15 @@
-        ENTITY_ICON: "mdi:checkbox-marked-circle-outline",
-        ENTITY_CLASS: None,
+        ENTITY_ICON: None,
+        ENTITY_CLASS: DEVICE_CLASS_SAFETY,
```
I suggest we revert these changes for now since it's not correct for these regular sensors.

Mariusthvdb replied:
done

Mariusthvdb committed:
reverted non binary status sensors

to use: "mdi:checkbox-marked-circle-outline" and set None to device_class

Mariusthvdb commented:
anyone on this failing test? was alright before.

MartinHjelmare commented:
I restarted CI and it passed now.

Quentame reviewed (approved):
Thanks!
Will check sensors to be binary or not, with a proper state.

Mariusthvdb commented:
<img width="94" alt="Schermafbeelding 2020-09-23 om 10 35 24" src="https://user-images.githubusercontent.com/33354141/93988210-bc9ce580-fd88-11ea-9571-972fb448e590.png">

seems the icon change for device_class on this integration didnt merge correctly? How to proceed here?

Quentame commented:
Your dev was for 0.116.0, wait the next release :wink:

Mariusthvdb commented:
ok thanks, sorry for the follow-up post. At least the device_class is now set, which is even more important.
